rfc/0017-policy-server-reuse-wasm-optimized-modules.md

```diff
@@ -0,0 +1,294 @@
+6. Once all the modules have been downloaded and optimized exit
+
+The typical Policy Server deployment is made by two Pods. In this case, one Pod
+will obtain the exclusive access to the shared lock. Its init container will
+eventually update the contents of the Persistent Volume and then it will start.
+The other Pod will be initially "stuck" with its init container waiting to acquire
+the shared lock. Once the lock is owned, the init container program will exit
+almost immediately, given all the modules and their optimized versions are
+already inside of the Persistent Volume.
+
+The usage of an init container has the following benefits:
+
+* Create a clear representation of the bootstrap process of the Policy Server
+  Pod
+* Avoid further complexity to be added to the Policy Server codebase
+
+The "download & optimize" code would be provided by a dedicated binary that is
+executed by an init container.
+
+To keep things simple, this binary can be stored inside of the already existing
+Policy Server container image. By doing that we do not complicate our build pipeline
+and we do not introduce a new container image that has to be mirrored when
+the Kubewarden stack is deployed inside of an air-gapped environment.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Usage of a Persistent Volume
+
+A Persistent Volume must be created. Our helm chart should not create the Persistent
+Volume, but rather rely on a [Storage Class](https://kubernetes.io/docs/concepts/storage/storage-classes/)
+to have that created.
+
+Development/testing environments should not be affected by that because
+k3s, minikube, kind and other similar projects provide a default Storage Class
+that is ready to be used.
+
+The majority of production environments already feature a Storage Class that
+can be reused.
+
+We can also change the helm chart to handle these possible configurations:
+
+* Use the default Storage Class
+* Use a Storage Class specified by the user
+* Do not use a Storage Class, use a Persistent Volume already defined by the
+  cluster administrator
+
+## Rust and Kubernetes Lease handling
+
+The complexity of building a lock system using the Kubernetes Lease resource is
+entirely inside of the Kubernetes client being used.
+
+In the Rust ecosystem the `kubert` crate provides this functionality. Despite being
+used by linkerd, the crate development/maintenance seems to be low.
+
+We can reduce the impact of this dependency by creating a dedicated program that
+performs the download and optimization task. If `kubert` is abandoned we
+can write a Go program that takes the lock and then `exec` the "download & optimize"
+Rust code (which would no longer handle the lock aspects).
+
+## Garbage collector
+
+Not really a drawback, but something to keep in mind. Over the time the size of
+the Persistent Volume will grow. We will have to introduce a way to ensure old
+modules (and their optimized counterparts) are removed when no longer needed.
```

Comment:
Maybe, we can define a cronjob for that. Using the same binary used to optimize. But with different command line flag to clean the PV.